packages/tldraw/src/lib/defaultExternalContentHandlers.ts

```diff
@@ -90,24 +92,49 @@ export function registerDefaultExternalContentHandlers(
 		}
 
 		const assetId: TLAssetId = AssetRecordType.createId(hash)
-
-		const asset = AssetRecordType.create({
+		const assetInfo = {
 			id: assetId,
 			type: isImageType ? 'image' : 'video',
 			typeName: 'asset',
 			props: {
 				name,
-				src: await FileHelpers.blobToDataUrl(file),
+				src: '',
 				w: size.w,
 				h: size.h,
 				mimeType: file.type,
 				isAnimated,
 			},
-		})
+		} as TLAsset
+
+		if (editor.hasExternalAssetHandler('blob')) {
+			assetInfo.props.src = assetId
+			const asset = await editor.getAssetForExternalContent({
+				type: 'blob',
+				assetInfo,
+				blob: file,
+			})
+
+			return asset!
+		}
 
-		return asset
+		assetInfo.props.src = await FileHelpers.blobToDataUrl(file)
+		return AssetRecordType.create(assetInfo)
 	})
 
+	if (persistenceKey) {
+		editor.registerExternalAssetHandler('blob', async ({ assetInfo, blob }) => {
```

Comment:
ok so i don't know if i can do a good job explaining this in a comment - half tempted to sketch what i mean in code, but also happy to try pairing.

in order to plumb this in, we've had to add a special new external asset handler, that's only ever called from one of _our_ external asset handlers. This is probably something I think we should avoid - I think these asset handlers are already pretty confusing, and adding another type to support this seems unnecesary.

The thing that I've been imagining here is that in the Tldraw component, we have some code that looks like this:
```ts
let onMount = props.onMount
if (props.persistenceKey) {
    onMount = (editor) => {
      editor.registerExternalAssetHandler('file', () => {
        // do the stuff from the default handler
        // but also do the blobby stuff
      })
      return props.onMount?.(editor)
    }
}

return <WhateverIsNext {...props} onMount={onMount} />
```

(i also imagined plumbing in the resolve asset stuff the same way - so rather than having `persistenceKey` go deep into these fns, it sort of stops at that component level)

I think this would help keep the APIs around these handlers a bit easier to conceptualise & understand, but I appreciate that this is something we've been plugging away at for ages now - if you wanted to land this as is then that's all good too

Reply:
No worries and pushing this to get better :) Definitely wanna get this right. 

I _think_ I kinda understand what you mean but maybe I'm confused on how onMount comes into play here (because it feels like onMount might be too late for the initial load? or maybe I'm wrong).

Yeah, I think it might be worth a pairing. I'll grab you when you have time!

Reply:
talked offline with Alex about this and simplified 👍